### ISSUE TITLE
Centralized monitoring/analytics service

### ISSUE THREAD

**Issue by gioscarda** [GeoNode/geonode#5051]
**Is your feature request related to a problem? Please describe.**
We want to have a centralized management of analytics data coming from one or more GeoNode instances. So we have to extend the current monitoring app in order to make available sending messages to a "Centralized Server" which comes with Logstash.
Each instance will be identified through the HOSTNAME and the IP.
Data, gzip compressed, will be sent at regular intervals (1 hour by default) via TCP.
Data to be sent:
- total number of requests
- total number of unique sessions,
- total number of unique users,
- total number of registered users at the end time,
- total number of layers at the end time,
- total number of documents at the end time,
- total number of maps at the end time,
- total number of errors
- for each resource:
    - resource type,
    - resource name,
    - resource URL,
    - total number of requests about this resource,
    - total number of unique sessions about this resource,
    - total number of unique users about this resource,
    - total number of resource downloads,
    - total number of OGC service requests about this resource,
    - total number of publication events
- for each country:
    - total number of requests about the country
- for each UA family:
    - total number of requests about the UA family

The whole service should be configurable from the admin interface.

**Describe the solution you'd like**
We can use the [python-logstash-async](https://pypi.org/project/python-logstash-async/) library which works asynchronously collecting log events from Python’s logging subsystem and then transmitting the collected events in a separate worker thread to Logstash. This way, the main application (or thread) where the log event occurred, doesn’t need to wait until the submission to the remote Logstash instance succeeded.
Metrics data, aggregated based on the time interval defined, should be retrieved using the current collection system and manipulated in order to send relevant information only.
These messages will be sent each time a scheduled celery task trigger the delivering process.

**Describe alternatives you've considered**
N.d.

**Additional context**
It implicitly implies a server side configuration with [Logstash](https://www.elastic.co/products/logstash).

**Comment by afabiani** [comment]
@gioscarda can you convert this to a GNIP?

Remember to update also this page https://github.com/GeoNode/geonode/wiki/GeoNode-Improvement-Proposals

**Comment by blewis** [comment]
This looks great.  Would this support WMS/WFS sharing between geonode instances?

**Comment by timlinux** [comment]
Great idea! +1 from me

**Comment by capooti** [comment]
+1 thanks!

**Comment by t-book** [comment]
Hi @gioscarda thanks a lot for this well-described PR. My +1. 

One question. Do you see any other option instead of storing the queue in an SQLite database? Maybe some Broker like Rabbitmq?  I must admit I do not know a lot of the monitoring app but having issues with SQLite in production over the last years.

**Comment by gioscarda** [comment]
Hi Ben,
not sure I understood your question, but this GNIP is only focused on monitoring resources which come from one or more GeoNode instances, not on sharing them.


Information about OGC hits, such as WMS and WFS, will be sent to the server so that you could visualize (through a Kibana visualization for example) these data for multiple instances in a single chart.



Regards,
Giovanni Scardapane
==
GeoServer Professional Services from the experts! Visit http://goo.gl/it488V for more information.
==
 
Giovanni Scardapane
Software Developer
 
GeoSolutions S.A.S.
Via di Montramito 3/A
55054  Massarosa (LU)
Italy
phone: +39 0584 962313
fax:      +39 0584 1660272
 
http://www.geo-solutions.it
http://twitter.com/geosolutions_it
 
-------------------------------------------------------
 
Con riferimento alla normativa sul trattamento dei dati personali (Reg. UE 2016/679 - Regolamento generale sulla protezione dei dati “GDPR”), si precisa che ogni circostanza inerente alla presente email (il suo contenuto, gli eventuali allegati, etc.) è un dato la cui conoscenza è riservata al/i solo/i destinatario/i indicati dallo scrivente. Se il messaggio Le è giunto per errore, è tenuta/o a cancellarlo, ogni altra operazione è illecita. Le sarei comunque grato se potesse darmene notizia.
 
This email is intended only for the person or entity to which it is addressed and may contain information that is privileged, confidential or otherwise protected from disclosure. We remind that - as provided by European Regulation 2016/679 “GDPR” - copying, dissemination or use of this e-mail or the information herein by anyone other than the intended recipient is prohibited. If you have received this email by mistake, please notify us immediately by telephone or e-mail.

**Comment by gannebamm** [comment]
@gioscarda 
I do not understand where this is tightly coupled with GeoNode, yet. The current monitoring App does provide an API to fetch the data you would like to ingest into the proposed _external_ (?) service?

This external service would not be part of GeoNode as far as I understand, but more like a piece of software you could configure to monitor GeoNode(s)?

So, long story short: Where will this PRs code get integrated into the GeoNode codebase?

BTW. I like the idea and concept. I would like to see this as one additional docker container in the next docker-composition, like rabbitmq, rclone and pg_dumper.

**Comment by afabiani** [comment]
@gannebamm I see your point. To be honest I would prefer this module to be a "contrib-app" instead of part of the core. In fact, it requires, at this stage at least, the installation of several components different from GeoNode.

**Comment by gioscarda** [comment]
Hi @t-book

> One question. Do you see any other option instead of storing the queue in an SQLite database? Maybe some Broker like Rabbitmq?

Honestly I have not examined the possibility of using Rabbitmq to queue messages. 
I focused my investigations on using the [python-logstash-async](https://pypi.org/project/python-logstash-async/) to have ready-to-use system for logging messages to logstash. It allows to possible alternatives to cache log events between emitting and transmission to the Logstash server:
- SQLite db (log events are cached even across process restarts and crashes)
- in memory cache (cached events will not be kept across process restarts)

It will be configurable from the django admin.

I could add the Rabbitmq usage in the "Future evolution" section.

**Comment by t-book** [comment]
> I could add the Rabbitmq usage in the "Future evolution" section.

As long as we can be sure that SQLite is well tested in this context f.e. thread safety I'm fine ;)

**Comment by gioscarda** [comment]
@gannebamm
> So, long story short: Where will this PRs code get integrated into the GeoNode codebase?

As suggested by @afabiani it will be a contrib app, [here](https://github.com/GeoNode/geonode-contribs/pull/7) the PR.

On GeoNode side I just need to introduce `django_celery_beat` in order to manage periodic tasks from admin UI and make the contrib app working.

I will also update the [django-contrib_apps](http://docs.geonode.org/en/master/advanced/index.html#geonode-django-contrib-apps) docs section.

Here the [GeoNode PR](https://github.com/GeoNode/geonode/pull/5066).

**Comment by capooti** [comment]
@gioscarda keep in mind that celery is overkill if you just need it for scheduling a task. A django command in crontab can be a better fit

**Comment by gioscarda** [comment]
@capooti I opted for a celery task to make the whole service configurable from the Django admin interface, without any system-level operations.

**Comment by capooti** [comment]
@gioscarda agreed, just make sure to have a fallback option if the system admin doesn't want to rely on celery.
